subsys/bluetooth/host_extensions/conn_evt_cb.c

```diff
@@ -0,0 +1,207 @@
+
+static nrfx_egu_t egu_inst = NRFX_EGU_INSTANCE(EGU_INST_IDX);
+
+struct conn_evt_cb_cfg {
+	struct bt_conn_evt_cb cb;
+	struct bt_conn *conn;
+	void *user_data;
+	uint32_t prepare_time_us;
+
+	uint8_t ppi_channel;
+	uint16_t conn_interval_us;
+};
+
+static struct conn_evt_cb_cfg cfg[CONFIG_BT_MAX_CONN];
+static struct k_work_delayable work_queue_items[CONFIG_BT_MAX_CONN];
+
+static void work_handler(struct k_work *work)
+{
+	uint8_t index = ARRAY_INDEX(work_queue_items, work);
+
+	cfg[index].cb.prepare(cfg[index].conn, cfg[index].user_data);
+
+	/* Schedule the callback in the connection event does not occur because
+	 * of peripheral latency or scheduling conflicts.
+	 */
+	k_work_reschedule(&work_queue_items[index], K_USEC(cfg[index].prepare_time_us));
+}
+
+static void egu_event_handler(uint8_t event_idx, void *context)
+{
+	ARG_UNUSED(context);
+
+	__ASSERT_NO_MSG(event_idx < ARRAY_SIZE(work_queue_items));
+	__ASSERT_NO_MSG(cfg[event_idx].conn_interval_us > cfg[event_idx].prepare_time_us);
+
+	const uint32_t timer_distance_us =
+		cfg[event_idx].conn_interval_us - cfg[event_idx].prepare_time_us;
+
+	k_work_reschedule(&work_queue_items[event_idx], K_USEC(timer_distance_us));
+}
+
+static void on_conn_param_updated(struct bt_conn *conn,
+				  uint16_t interval,
+				  uint16_t latency,
+				  uint16_t timeout)
+{
+	uint8_t index = bt_conn_index(conn);
+
+	if (!cfg[index].cb.prepare) {
+		return;
+	}
+
+	cfg[index].conn_interval_us = BT_CONN_INTERVAL_TO_US(interval);
+}
+
+static void on_disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	uint8_t index = bt_conn_index(conn);
+
+	if (!cfg[index].cb.prepare) {
+		return;
+	}
+
+	cfg[index].cb.prepare = NULL;
+	nrfx_egu_int_disable(&egu_inst, nrf_egu_channel_int_get(index));
+	(void)k_work_cancel_delayable(&work_queue_items[index]);
+
+	nrfx_err_t err = nrfx_gppi_channel_free(cfg[index].ppi_channel);
+
+	__ASSERT_NO_MSG(err == NRFX_SUCCESS);
+}
+
+BT_CONN_CB_DEFINE(conn_params_updated_cb) = {
+	.le_param_updated = on_conn_param_updated,
+	.disconnected = on_disconnected,
+};
+
+static int setup_connection_event_trigger(struct conn_evt_cb_cfg *cfg, uint8_t index)
+{
+	int err;
+	uint16_t conn_handle;
+
+	err = bt_hci_get_conn_handle(cfg->conn, &conn_handle);
+	if (err) {
+		return err;
+	}
+
+	sdc_hci_cmd_vs_get_next_conn_event_counter_t get_next_conn_event_count_params = {
+		.conn_handle = conn_handle,
+	};
+	sdc_hci_cmd_vs_get_next_conn_event_counter_return_t get_next_conn_event_count_ret_params;
+
+	err = hci_vs_sdc_get_next_conn_event_counter(
+		&get_next_conn_event_count_params, &get_next_conn_event_count_ret_params);
+	if (err) {
+		return err;
+	}
+
+	const sdc_hci_cmd_vs_set_conn_event_trigger_t params = {
+		.conn_handle = conn_handle,
+		.role = SDC_HCI_VS_CONN_EVENT_TRIGGER_ROLE_CONN,
+		.ppi_ch_id = cfg->ppi_channel,
+		.period_in_events = 1, /* Trigger every connection event. */
+		.conn_evt_counter_start =
+			get_next_conn_event_count_ret_params.next_conn_event_counter + 10,
```

Comment:
Why the `+ 10`?

Reply:
It is added just to ensure we don't start it too late (which is not allowed). We plan to replace it with a new HCI command which doesn't have this logic

Reply:
Ok. I was hoping that if we just set it to zero it would trigger on the next event.

I note that depending on connection interval, 10 connection events is a noticeable delay for a e.g. a mouse to start reporting after reconnecting. This could be critical to make a 'deep sleep until accelerometer input' feel responsive to an end user.